### PR TITLE
DemonstrationRecorder: Add numStepsToRecord feature.

### PR DESCRIPTION
**Pull request by Hunter-Unity**
### Proposed change(s)

Updates the DemonstrationRecorder so that you can specify the number of steps to record. When this number is reached, the editor will stop and save the demo file.

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs [hunk 1]
```diff
@@ -32,6 +32,11 @@ public class DemonstrationRecorder : MonoBehaviour
         [Tooltip("Whether or not to record demonstrations.")]
         public bool Record;
 
+        /// <summary>
+        /// Number of steps to record. Editor will stop playing when it hits this threshold.
+        /// </summary>
+        public int numStepsToRecord = 5000;
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:38]:
```suggestion
        public int NumStepsToRecord = 5000;
```
We're using PascalCase for public fields and properties.

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs [hunk 2]
```diff
@@ -32,6 +32,11 @@ public class DemonstrationRecorder : MonoBehaviour
         [Tooltip("Whether or not to record demonstrations.")]
         public bool Record;
 
+        /// <summary>
+        /// Number of steps to record. Editor will stop playing when it hits this threshold.
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:36]: Can you add this as a `[Tooltip(...)]` too?

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs [hunk 3]
```diff
@@ -70,6 +75,11 @@ void Update()
             {
                 LazyInitialize();
             }
+
+            if (m_DemoWriter.NumSteps >= numStepsToRecord)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:79]: Should we only quit if `NumStepsToRecord > 0` too?

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs [hunk 4]
```diff
@@ -70,6 +75,11 @@ void Update()
...
+            {
+                UnityEditor.EditorApplication.isPlaying = false;
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:81]:
```suggestion
                Application.Quit(0);
#if UNITY_EDITOR
                UnityEditor.EditorApplication.isPlaying = false;
#endif
```
This is what's causing the yamato failures for the standalone build. And in that case, we want to exit too.

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs (hunk 1 above)
Comment (ervteng) [inline com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:38]: Is there a way to specify infinite steps (i.e. current behavior?)

Comment (Hunter-Unity) [reply com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:38]: no, i figured if you want more than 5k you could just change the NumStepsToRecord. i'm not sure what the typical range is for the number of steps for demos.

Comment (Hunter-Unity) [reply com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs]: changed to PascalCase

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs (hunk 2 above)
Comment (Hunter-Unity) [reply com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:36]: added tooltip

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs (hunk 3 above)
Comment (Hunter-Unity) [reply com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs]: added >0 check. good idea

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs (hunk 4 above)
Comment (Hunter-Unity) [reply com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:84]: ahhh of course. added suggested changes.

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs [hunk 5]
```diff
@@ -70,6 +76,14 @@ void Update()
             {
                 LazyInitialize();
             }
+
+            if (m_DemoWriter.NumSteps > 0 && m_DemoWriter.NumSteps  >= NumStepsToRecord)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:80]:
```suggestion
            if (NumStepsToRecord > 0 && m_DemoWriter.NumSteps  >= NumStepsToRecord)
```
Not sure the first part will accomplish anything as it's written now.

Comment (Hunter-Unity) [reply com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs]: yeah caught that as well. it's fixed now.

File: com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs [hunk 6]
```diff
@@ -32,6 +32,14 @@ public class DemonstrationRecorder : MonoBehaviour
         [Tooltip("Whether or not to record demonstrations.")]
         public bool Record;
 
+        /// <summary>
+        /// Number of steps to record. If this number is higher than zero, the editor will stop
+        /// playing when it reaches this threshold.
+        /// </summary>
+        [Tooltip("Number of steps to record. If this number is higher than zero, the editor will stop " +
```
Comment (ervteng) [inline com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs:39]: Maybe change to "Number of steps to record. The editor will stop playing when it reaches this threshold. Set to zero to record indefinitely."

Comment (Hunter-Unity) [reply com.unity.ml-agents/Runtime/Demonstrations/DemonstrationRecorder.cs]: updated to use this language ^

**Pull request merged**